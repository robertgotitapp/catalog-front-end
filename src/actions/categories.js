import { CategoriesAction } from '../utils/const';
import { post, get } from '../utils/requests';

export function addCategory(accessToken, category) {
  return {
    type: CategoriesAction.ADD_CATEGORY,
    promise: post('http://127.0.0.1:5000/categories', accessToken, category),
  };
}

export function getCategories(offset, limit) {
  return {
    type: CategoriesAction.GET_CATEGORIES,
    promise: get(`http://127.0.0.1:5000/categories?offset=${offset}&limit=${limit}`),
  };
}

export function selectCurrentCategory(selectedCategory) {
  return {
    type: CategoriesAction.SELECT_CURRENT_CATEGORY,
    payload: { currentCategory: selectedCategory },
  };
}
