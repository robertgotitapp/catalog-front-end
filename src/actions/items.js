import { ItemsAction, HeadersType } from '../utils/const';
import {
  get, post, update, remove,
} from '../utils/requests';

export function addItem(categoryId, item) {
  return {
    type: ItemsAction.ADD_ITEM,
    promise: post(`http://127.0.0.1:5000/categories/${categoryId}/items`,
      [HeadersType.CONTENTTYPE, HeadersType.AUTHORIZATION],
      item),
  };
}

export function getItems(categoryId, offset, limit) {
  return {
    type: ItemsAction.GET_ITEMS,
    promise: get(`http://127.0.0.1:5000/categories/${categoryId}/items?offset=${offset}&limit=${limit}`),
  };
}

export function removeItem(categoryId, itemId) {
  return {
    type: ItemsAction.DELETE_ITEM,
    promise: remove(`http://127.0.0.1:5000/categories/${categoryId}/items/${itemId}`,
      [HeadersType.CONTENTTYPE, HeadersType.AUTHORIZATION]),
  };
}

export function updateItem(categoryId, itemId, item) {
  console.log(categoryId);
  console.log(item);
  return {
    type: ItemsAction.UPDATE_ITEM,
    promise: update(
      `http://127.0.0.1:5000/categories/${categoryId}/items/${itemId}`,
      [HeadersType.CONTENTTYPE, HeadersType.AUTHORIZATION],
      item,
    ),
  };
}
