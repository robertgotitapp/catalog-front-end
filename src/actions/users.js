export const ADD_USER = 'ADD_USER'
export const AUTHENTICATE_USER = 'AUTHENTICATE_USER'

export function addUser(user) {
    return {
        type: ADD_USER,
        user
    }
}

export function handleAddUser(user) {
    return (dispatch) => {
        dispatch(addUser(user))
    }
}

