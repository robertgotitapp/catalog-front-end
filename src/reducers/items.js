import { ItemsAction } from '../utils/const';

const initialState = { items: {}, currentPage: null };

export default function items(state = initialState, action) {
  switch (action.type) {
    case ItemsAction.GET_ITEMS_SUCCESS:
      return {
        ...state,
        items: action.payload.items,
        totalItems: action.payload.total_items,
      };
    case ItemsAction.ADD_ITEM_SUCCESS:
      return {
        ...state,
        [action.payload.id]: action.payload,
      };
    case ItemsAction.UPDATE_ITEM_SUCCESS:
      return {
        ...state,
        [action.payload.id]: action.payload,
      };
    case ItemsAction.SELECT_ITEM_PAGE:
      return {
        ...state,
        currentPage: action.payload.currentPage,
      };
    default:
      return state;
  }
}
