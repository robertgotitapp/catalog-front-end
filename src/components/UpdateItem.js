import React, { Component } from 'react';
import { connect } from 'react-redux';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import { updateItem } from '../actions/items';

export class UpdateItem extends Component {
    state = {
      name: this.props.item.name,
      price: this.props.item.price,
      description: this.props.item.description,
    }

    handleChange = (e) => {
      this.setState({
        [e.target.name]: e.target.value,
      });
    }

    handleSubmit = (e) => {
      e.preventDefault();
      this.props.updateItem(this.props.item.category_id, this.props.item.id, { ...this.state });
    }

    render() {
      const { name, price, description } = this.state;

      return (
        <div>
          <Form onSubmit={this.handleSubmit}>
            <Form.Group>
              <Form.Label>Item Name</Form.Label>
              <Form.Control type="text" name="name" onChange={this.handleChange} value={name} />
            </Form.Group>
            <Form.Group>
              <Form.Label>Item Price</Form.Label>
              <Form.Control type="number" name="price" onChange={this.handleChange} value={price} />
            </Form.Group>
            <Form.Group>
              <Form.Label>Item Description</Form.Label>
              <Form.Control
                type="text"
                name="description"
                onChange={this.handleChange}
                value={description}
              />
            </Form.Group>
            <Button variant="primary" type="submit">
              Submit
            </Button>
          </Form>
        </div>
      );
    }
}

function mapStateToProps({ itemsReducer }, { match }) {
  const itemId = Number(match.params.id);
  const selectedItem = Object.values(itemsReducer.items)
    .find(item => item.id === itemId);
  return {
    item: selectedItem,
  };
}

const mapDispatchtoProps = {
  updateItem,
};

export default connect(mapStateToProps, mapDispatchtoProps)(UpdateItem);
