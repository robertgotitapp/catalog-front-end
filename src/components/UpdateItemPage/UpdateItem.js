import React, { Component } from 'react';
import { connect } from 'react-redux';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import Alert from 'react-bootstrap/Alert';
import { updateItem, getItems, selectItemPage } from '../../actions/items';
import { PaginationConfig } from '../../utils/const';

export class UpdateItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: this.props.item.name,
      price: this.props.item.price,
      description: this.props.item.description,
      alerts: '',
    };
  }

    validateInput = () => {
      const {
        name, description, price,
      } = this.state;
      const errors = {};
      if (name.length <= 5) {
        errors.name = 'Name must be longer than 5 characters.';
      }
      if (description.length > 200) {
        errors.description = 'Description must be within 200 characters.';
      }
      if (price <= 0) {
        errors.price = 'Price must be positive number.';
      }
      if (Object.keys(errors).length !== 0) {
        this.setState({
          alerts: errors,
        });
        return false;
      }
      return true;
    }

    handleChange = (e) => {
      this.setState({
        [e.target.name]: e.target.value,
      });
    }

    handleSubmit = (e) => {
      e.preventDefault();
      const { name, price, description } = this.state;
      if (this.validateInput()) {
        this.props.updateItem(
          this.props.item.category_id,
          this.props.item.id,
          { name, price, description },
        )
          .then((res) => {
            if (res.statusCode) {
              this.props.selectItemPage(PaginationConfig.DEFAULT_PAGE);
              const limit = PaginationConfig.ITEMS_PER_PAGE;
              this.props.getItems(this.props.currentCategory,
                PaginationConfig.DEFAULT_OFFSET, limit);
              this.props.history.push('/');
            } else {
              this.setState({
                alerts: res.errors.message,
              });
            }
          });
      }
    }

    render() {
      const {
        name, price, description,
      } = this.state;

      return (
        <div>
          <h2>Update Item</h2>
          <Form onSubmit={this.handleSubmit}>
            <Form.Group>
              <Form.Label>Item Name</Form.Label>
              <Form.Control type="text" name="name" onChange={this.handleChange} value={name} />
            </Form.Group>
            <Form.Group>
              <Form.Label>Item Price</Form.Label>
              <Form.Control type="number" name="price" onChange={this.handleChange} value={price} />
            </Form.Group>
            <Form.Group>
              <Form.Label>Item Description</Form.Label>
              <Form.Control
                type="text"
                name="description"
                onChange={this.handleChange}
                value={description}
              />
            </Form.Group>
            <Button variant="primary" type="submit">
              Submit
            </Button>
          </Form>
          {
            // Loop through alerts dictionary and display an alert for each
            // of one item in dictionary
            this.state.alerts
            && Object.keys(this.state.alerts).map((key, id) => (
              <Alert key={id} variant='danger'>
                {key}
                :
                {this.state.alerts[key]}
              </Alert>
            ))
          }
        </div>
      );
    }
}

function mapStateToProps({ items, categories }, { match }) {
  const itemId = Number(match.params.id);
  const selectedItem = Object.values(items.items)
    .find(item => item.id === itemId);
  return {
    item: selectedItem,
    currentCategory: categories.currentCategory,
  };
}

const mapDispatchtoProps = {
  updateItem,
  getItems,
  selectItemPage,
};

export default connect(mapStateToProps, mapDispatchtoProps)(UpdateItem);
