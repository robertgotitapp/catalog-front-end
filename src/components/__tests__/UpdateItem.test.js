import React from 'react';
import { shallow } from 'enzyme';
import Form from 'react-bootstrap/Form';
import Alert from 'react-bootstrap/Alert';
import { UpdateItem } from '../UpdateItem';
import { updateItem } from '../../actions/items';

describe('components/UpdateItem', () => {
  const props = {
    item: {
      category_id: 1,
      created: '2019-06-17T13:49:00+00:00',
      description: 'Item 4',
      id: 11,
      name: 'Item 4',
      price: 19,
      updated: '2019-06-17T13:49:00+00:00',
    },
    currentCategory: {
      created: '2019-06-17T13:46:18+00:00',
      description: 'Laptop',
      id: 1,
      name: 'Laptop',
      updated: '2019-06-17T13:46:18+00:00',
    },
    updateItem,
  };
  const container = shallow(<UpdateItem {...props} />);

  it('should render correctly', () => {
    expect(container).toMatchSnapshot();
  });

  it('should contain exact number of item name', () => {
    expect(container.find('[name="name"]')).toHaveLength(1);
  });

  it('should contain exact number of item price', () => {
    expect(container.find('[name="price"]')).toHaveLength(1);
  });

  it('should contain exact number of item description', () => {
    expect(container.find('[name="description"]')).toHaveLength(1);
  });

  it('should trigger state change when FormControl name got input', () => {
    const itemNameInput = container.find('[name="name"]');
    itemNameInput.simulate('change', { target: { name: 'name', value: 'item 1' } });
    expect(container.state().name).toEqual('item 1');
  });

  it('should trigger state change when FormControl description got input', () => {
    const itemDescriptionInput = container.find('[name="description"]');
    itemDescriptionInput.simulate('change', { target: { name: 'description', value: 'description' } });
    expect(container.state().description).toEqual('description');
  });

  it('should trigger state change when FormControl price got input', () => {
    const itemPriceInput = container.find('[name="price"]');
    itemPriceInput.simulate('change', { target: { name: 'price', value: 100 } });
    expect(container.state().price).toEqual(100);
  });

  it('should trigger handleSubmit handler when form is submitted', () => {
    container.instance().handleSubmit = jest.fn();
    container.instance().forceUpdate();
    const form = container.find(Form);
    form.simulate('submit');
    expect(container.instance().handleSubmit).toHaveBeenCalled();
  });

  it('should trigger alert when there is error message', () => {
    container.setState({
      name: 'Item 4',
      price: 19,
      description: 'Item 4',
      toHome: false,
      alerts: {
        name: 'Name has to be at least 5 characters length',
      },
    });
    expect(container.find(Alert)).toHaveLength(1);
  });
});
