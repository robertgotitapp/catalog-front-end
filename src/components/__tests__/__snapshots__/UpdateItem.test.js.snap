// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/UpdateItem should render correctly 1`] = `
<div>
  <h2>
    Update Item
  </h2>
  <Form
    inline={false}
    onSubmit={[Function]}
  >
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Item Name
      </FormLabel>
      <FormControl
        name="name"
        onChange={[Function]}
        type="text"
        value="Item 4"
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Item Price
      </FormLabel>
      <FormControl
        name="price"
        onChange={[Function]}
        type="number"
        value={19}
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Item Description
      </FormLabel>
      <FormControl
        name="description"
        onChange={[Function]}
        type="text"
        value="Item 4"
      />
    </FormGroup>
    <Button
      active={false}
      disabled={false}
      type="submit"
      variant="primary"
    >
      Submit
    </Button>
  </Form>
</div>
`;
